lib/daru/category.rb

```diff
@@ -0,0 +1,107 @@
+module Daru
+  module Category
+    attr_accessor :coding_scheme, :base_category
+
+    def initialize_catogory data, opts={}
+      @type = :category
+
+      # Create a hash to map each category to positional indexes
+      categories = data.each_with_index.group_by(&:first)
+      @cat_hash = categories.map { |cat, group| [cat, group.map(&:last)] }.to_h
+
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = categories.keys.each_with_index.to_h
+
+      # Inverse mapping of map_cat_int
+      @map_int_cat = map_cat_int.invert
+
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = map_cat_int.values_at(*data)
+
+      # Specify if the categories are ordered or not.
+      # By default its unordered
+      @order = opts[:order] || false
+
+      # The coding scheme to code with. Default is dummy coding.
+      @coding_scheme = :dummy
+
+      # Base category which won't be present in the coding
+      @base_category = @cat_hash.keys.first
+
+      # Stores the name of the vector
+      @name = opts[:name]
+    end
+
+    def each
+      return enum_for(:each) unless block_given?
+      @array.each { |pos| yield @map_int_cat[pos] }
+      self
+    end
+
+    def to_a
+      each.to_a
+    end
+
+    def size
+      @array.size
+    end
+
+    def ordered?
+      @order
+    end
+
+    def category
```

Comment:
Maybe this should be called `categories` in plural? Because it returns an array of all available categories rather than just one category.